lib/stm32/l4/rcc.c

```diff
@@ -341,6 +380,83 @@ uint32_t rcc_system_clock_source(void)
 	return (RCC_CFGR >> RCC_CFGR_SWS_SHIFT) & RCC_CFGR_SWS_MASK;
 }
 
+/**
+ * Setup clocks to run from PLL.
+ *
+ * The arguments provide the pll source, multipliers, dividers, all that's
+ * needed to establish a system clock.
+ *
+ * @param clock clock information structure.
+ */
+void rcc_clock_setup_pll(const struct rcc_clock_scale *clock)
+{
+  /* Enable internal high-speed oscillator (HSI16). */
+  rcc_osc_on(RCC_HSI16);
+  rcc_wait_for_osc_ready(RCC_HSI16);
+  
+  /* Select HSI16 as SYSCLK source. */
+  rcc_set_sysclk_source(RCC_PLLCFGR_PLLSRC_HSI16);
+  
+  /* Enable external high-speed oscillator (HSE). */
+  if (clock->pll_source == RCC_PLLCFGR_PLLSRC_HSE) {
+    rcc_osc_on(RCC_HSE);
+    rcc_wait_for_osc_ready(RCC_HSE);
+  }
+  
+  /* Set the VOS scale mode */
+  rcc_periph_clock_enable(RCC_PWR);
+  pwr_set_vos_scale(clock->voltage_scale);
+  
+  /*
+   * Set prescalers for AHB, ADC, APB1, APB2.
+   * Do this before touching the PLL (TODO: why?).
+   */
+  rcc_set_hpre(clock->hpre);
+  rcc_set_ppre1(clock->ppre1);
+  rcc_set_ppre2(clock->ppre2);
+  
+  /* Disable PLL oscillator before changing its configuration. */
+  rcc_osc_off(RCC_PLL);
+  
+  /* Configure the PLL oscillator. */
+  rcc_set_main_pll(clock->pll_source, clock->pllm, clock->plln,
+                         clock->pllp, clock->pllq, clock->pllr);
+  
```

Comment:
all whitespace here is off as well, you need to use the normal hard tab styles please